File: android/src/main/java/co/techFlow/auto_start_flutter/AutoStartFlutterPlugin.java
package co.techFlow.auto_start_flutter;

import androidx.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import io.flutter.embedding.android.FlutterActivity;
import io.flutter.embedding.engine.FlutterEngine;
import io.flutter.plugin.common.BinaryMessenger;
import android.provider.Settings;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.ComponentName;
import java.util.*;
import android.content.pm.ResolveInfo;
import android.util.Log;

/** AutoStartFlutterPlugin */
public class AutoStartFlutterPlugin implements FlutterPlugin, MethodCallHandler {

  private MethodChannel channel;
  private Context context;


  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    onAttach(flutterPluginBinding.getApplicationContext(),flutterPluginBinding.getBinaryMessenger());
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "com.techflow.co/auto_start_flutter");
    channel.setMethodCallHandler(this);
  }

  private void onAttach(Context applicationContext, BinaryMessenger messenger) {
    this.context = applicationContext;
    channel = new MethodChannel(messenger, "com.techflow.co/auto_start_flutter");
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if(call.method.equals("permit-auto-start")){
      addAutoStartup();
    } else  if (call.method.equals("isAutoStartPermission")) {
      String manufacturer = android.os.Build.MANUFACTURER;
      if ("xiaomi".equalsIgnoreCase(manufacturer)) {
        result.success(true);
      } else if ("poco".equalsIgnoreCase(manufacturer)) {
        result.success(true);
      } else if ("redmi".equalsIgnoreCase(manufacturer)) {
        result.success(true);
      } else if ("letv".equalsIgnoreCase(manufacturer)) {
        result.success(true);
      } else if ("oppo".equalsIgnoreCase(manufacturer)) {
        result.success(true);
      } else if ("vivo".equalsIgnoreCase(manufacturer)) {
        result.success(true);
      } else if ("letv".equalsIgnoreCase(manufacturer)) {
        result.success(true);
      } else if ("honor".equalsIgnoreCase(manufacturer)) {
        result.success(true);
      } else if ("huawei".equalsIgnoreCase(manufacturer)) {
        result.success(true);
      }else if ("samsung".equalsIgnoreCase(manufacturer)) {
        result.success(true);
      }else if ("oneplus".equalsIgnoreCase(manufacturer)) {
        result.success(true);
      }else if ("nokia".equalsIgnoreCase(manufacturer)) {
        result.success(true);
      }else if ("asus".equalsIgnoreCase(manufacturer)) {
        result.success(true);
      } else if ("realme".equalsIgnoreCase(manufacturer)) {
        result.success(true);
      }else{
        result.success(false);
      }
    }else {
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }


  private void addAutoStartup() {
    try {
      Intent intent = new Intent();
      intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
      String manufacturer = android.os.Build.MANUFACTURER;
      if ("xiaomi".equalsIgnoreCase(manufacturer)) {
        intent.setComponent(new ComponentName("com.miui.securitycenter", "com.miui.permcenter.autostart.AutoStartManagementActivity"));
      } else if ("poco".equalsIgnoreCase(manufacturer)) {
        intent.setComponent(new ComponentName("com.miui.securitycenter", "com.miui.permcenter.autostart.AutoStartManagementActivity"));
      } else if ("redmi".equalsIgnoreCase(manufacturer)) {
        intent.setComponent(new ComponentName("com.miui.securitycenter", "com.miui.permcenter.autostart.AutoStartManagementActivity"));
      } else if ("letv".equalsIgnoreCase(manufacturer)) {
        intent.setComponent(new ComponentName("com.letv.android.letvsafe", "com.letv.android.letvsafe.AutobootManageActivity"));
      } else if ("oppo".equalsIgnoreCase(manufacturer)) {
        intent.setComponent(new ComponentName("com.coloros.safecenter", "com.coloros.safecenter.permission.startup.StartupAppListActivity"));
      } else if ("vivo".equalsIgnoreCase(manufacturer)) {
        intent.setComponent(new ComponentName("com.vivo.permissionmanager", "com.vivo.permissionmanager.activity.BgStartUpManagerActivity"));
      } else if ("letv".equalsIgnoreCase(manufacturer)) {
        intent.setComponent(new ComponentName("com.letv.android.letvsafe", "com.letv.android.letvsafe.AutobootManageActivity"));
      } else if ("honor".equalsIgnoreCase(manufacturer)) {
        intent.setComponent(new ComponentName("com.huawei.systemmanager", "com.huawei.systemmanager.optimize.process.ProtectActivity"));
      } else if ("huawei".equalsIgnoreCase(manufacturer)) {
        intent.setComponent(new ComponentName("com.huawei.systemmanager", "com.huawei.systemmanager.startupmgr.ui.StartupNormalAppListActivity"));
      }else if ("samsung".equalsIgnoreCase(manufacturer)) {
        intent.setComponent(new ComponentName("com.samsung.android.lool", "com.samsung.android.sm.battery.ui.BatteryActivity"));
      }else if ("oneplus".equalsIgnoreCase(manufacturer)) {
        intent.setComponent(new ComponentName("com.oneplus.security", "com.oneplus.security.chainlaunch.view.ChainLaunchAppListActivity"));
      }else if ("nokia".equalsIgnoreCase(manufacturer)) {
        intent.setComponent(new ComponentName("com.evenwell.powersaving.g3", "com.evenwell.powersaving.g3.exception.PowerSaverExceptionActivity"));
      }else if ("asus".equalsIgnoreCase(manufacturer)) {
        intent.setComponent(new ComponentName("com.asus.mobilemanager", "com.asus.mobilemanager.autostart.AutoStartActivy"));
      } else if ("realme".equalsIgnoreCase(manufacturer)) {
        intent.setAction(Settings.ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS);
    }

      List<ResolveInfo> list = context.getPackageManager().queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY);
      if  (list.size() > 0) {
        context.startActivity(intent);
      }
    } catch (Exception e) {
      Log.e("exc" , String.valueOf(e));
    }
  }
}

File: android/src/main/kotlin/co/techFlow/auto_start_flutter/AutoStartFlutterPlugin.kt
package co.techFlow.auto_start_flutter

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.PowerManager
import android.provider.Settings
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import android.util.Log
import io.flutter.plugin.common.MethodChannel.Result


class AutoStartFlutterPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.techflow.co/auto_start_flutter")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "permit-auto-start" -> {
                val success = openAutoStartSettings()
                result.success(success)
            }
            "registerBootCallback" -> {
                val callbackHandle = call.argument<Long>("callbackHandle")
                if (callbackHandle != null) {
                    val prefs = context.getSharedPreferences(BootReceiver.SHARED_PREFS_KEY, Context.MODE_PRIVATE)
                    prefs.edit().putLong(BootReceiver.CACHED_CALLBACK_KEY, callbackHandle).apply()
                    result.success(true)
                } else {
                    result.error("INVALID_ARGUMENT", "callbackHandle must not be null", null)
                }
            }
            "isAutoStartPermission" -> {
                 val intents = AutoStartIntents.getIntents(context)
                 var found = false
                 for (intent in intents) {
                     try {
                         intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                         val list = context.packageManager.queryIntentActivities(
                             intent, PackageManager.MATCH_DEFAULT_ONLY
                         )
                         if (list.size > 0) {
                             found = true
                             break
                         }
                     } catch (e: Exception) {
                         // Ignore and try next intent
                     }
                 }
                 result.success(found)
             }
            "isBatteryOptimizationDisabled" -> {
                val packageName = context.packageName
                val pm = context.getSystemService(Context.POWER_SERVICE) as PowerManager
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    result.success(pm.isIgnoringBatteryOptimizations(packageName))
                } else {
                    result.success(true)
                }
            }
            "disableBatteryOptimization" -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    val intent = Intent(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS)
                    intent.data = Uri.parse("package:" + context.packageName)
                    intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                    context.startActivity(intent)
                }
                result.success(null)
            }
            "startForegroundService" -> {
                try {
                    val title = call.argument<String>("title")
                    val content = call.argument<String>("content")
                    val serviceIntent = Intent(context, KeepAliveService::class.java).apply {
                        putExtra("title", title)
                        putExtra("content", content)
                    }
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                        context.startForegroundService(serviceIntent)
                    } else {
                        context.startService(serviceIntent)
                    }
                    result.success(true)
                } catch (e: Exception) {
                    result.error("FOREGROUND_SERVICE_ERROR", e.message, null)
                }
            }
            "stopForegroundService" -> {
                try {
                    val serviceIntent = Intent(context, KeepAliveService::class.java)
                    context.stopService(serviceIntent)
                    result.success(true)
                } catch (e: Exception) {
                    result.error("FOREGROUND_SERVICE_ERROR", e.message, null)
                }
            }
            "openAppInfo" -> {
                try {
                    val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
                    intent.data = Uri.fromParts("package", context.packageName, null)
                    intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                    context.startActivity(intent)
                    result.success(true)
                } catch (e: Exception) {
                    result.success(false)
                }
            }
            "getDeviceManufacturer" -> {
                result.success(Build.MANUFACTURER)
            }
            "openCustomSetting" -> {
                val packageName = call.argument<String>("packageName")
                val activityName = call.argument<String>("activityName")
                if (packageName != null && activityName != null) {
                    try {
                        val intent = Intent()
                        intent.component = android.content.ComponentName(packageName, activityName)
                        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                        context.startActivity(intent)
                        result.success(true)
                    } catch (e: Exception) {
                        result.success(false)
                    }
                } else {
                    result.error("INVALID_ARGUMENT", "Package name and activity name must not be null", null)
                }
            }
            "scheduleTask" -> {
                val timestamp = call.argument<Long>("timestamp")
                val callbackHandle = call.argument<Long>("callbackHandle")
                val taskId = call.argument<String>("taskId")
                
                if (timestamp != null && callbackHandle != null && taskId != null) {
                    val success = scheduleAndroidAlarm(timestamp, callbackHandle, taskId)
                    result.success(success)
                } else {
                    result.error("INVALID_ARGUMENT", "timestamp, callbackHandle, and taskId must not be null", null)
                }
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun scheduleAndroidAlarm(timestamp: Long, callbackHandle: Long, taskId: String): Boolean {
        return try {
            val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as android.app.AlarmManager
            val intent = Intent(context, ScheduledTaskReceiver::class.java).apply {
                putExtra(ScheduledTaskReceiver.EXTRA_CALLBACK_HANDLE, callbackHandle)
                putExtra(ScheduledTaskReceiver.EXTRA_TASK_ID, taskId)
            }
            
            val pendingIntent = android.app.PendingIntent.getBroadcast(
                context,
                taskId.hashCode(),
                intent,
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    android.app.PendingIntent.FLAG_UPDATE_CURRENT or android.app.PendingIntent.FLAG_IMMUTABLE
                } else {
                    android.app.PendingIntent.FLAG_UPDATE_CURRENT
                }
            )
            
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                alarmManager.setExactAndAllowWhileIdle(android.app.AlarmManager.RTC_WAKEUP, timestamp, pendingIntent)
            } else {
                alarmManager.setExact(android.app.AlarmManager.RTC_WAKEUP, timestamp, pendingIntent)
            }
            true
        } catch (e: Exception) {
            Log.e("AutoStartPlugin", "Failed to schedule alarm: ${e.message}")
            false
        }
    }

    private fun openAutoStartSettings(): Boolean {
        val intents = AutoStartIntents.getIntents(context)
        for (intent in intents) {
            try {
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                val list = context.packageManager.queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY)
                if (list.size > 0) {
                    context.startActivity(intent)
                    return true
                }
            } catch (e: Exception) {
                // Ignore and try next intent
            }
        }
        return false
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}
